RUM Auto-instrumentation (#25947)
DataDog/documentation: content/en/real_user_monitoring/browser/setup/server.md

---
title: RUM Browser Monitoring Server Side Instrumentation
beta: true
code_lang: server
type: multi-code-lang
code_lang_weight: 2
further_reading:
- link: '/real_user_monitoring/explorer/'
  tag: 'Documentation'
  text: 'Learn about the RUM Explorer'
- link: '/logs/log_collection/javascript/'
  tag: 'Documentation'
  text: 'Learn about the Datadog Browser SDK for Logs'
---

{{< callout url="https://www.datadoghq.com/private-beta/rum-sdk-auto-injection/" btn_hidden="false" header="Access the Preview!">}}

> RUM SDK Injection in Preview. Sign up for access! You can also request early access if you are using Apache HTTPD, Tomcat, or another web server technology.

Reviewer: ```suggestion
RUM SDK Injection is in Preview. Sign up for access! You can also request early access if you are using Apache HTTPD, Tomcat, or another web server technology.
```

Author: This callout got removed entirely.

{{< /callout >}}

## Overview

Datadog RUM Server Side Instrumentation (SDK Injection) lets you opt into Real User Monitoring (RUM) automatically by instrumenting web applications served through a web server or proxy.

RUM Server Side Instrumentation works by injecting a RUM SDK JavaScript scriptlet into the HTML responses being served through a web server or proxy.

> After your applications have been instrumented, you can begin configuring your RUM application in Datadog.

Reviewer: ```suggestion
After your applications have been instrumented, you can configure your RUM application in Datadog.
```

## Prerequisites

The automatic installation method requires that you have the [Datadog Agent][2] installed.

> ## Setup your RUM application

Reviewer: ```suggestion
## Set up your RUM application
```

<div class="alert alert-warning">To request support for a web server that is not listed here, <a href="https://www.datadoghq.com/private-beta/rum-sdk-auto-injection/">fill out this form.</a></div>

{{< tabs >}}

> {{% tab "Nginx" %}}

Reviewer: I would style it as NGINX everywhere, since that's how NGINX itself does it, but I'm not sure if you're using existing RUM styles here. Just wanted to flag!

Author: Yes, I was torn on this. Seems like it's either NGINX or nginx, but to follow our existing styles, I decided to go with Nginx.

The Datadog RUM Server Side Instrumentation method leverages the [Nginx Dynamic Modules capability][1] to implement a response body filter. The filter injects the RUM SDK into the response body for responses identified as HTML.

To automatically instrument your RUM application:

1. In Datadog, navigate to the [**Digital Experience > Add an Application Page**][2] and select the JavaScript (JS) application type.
2. Select **Auto-Instrumentation** and **NGINX**.
3. Set your Session and Session Replay sample rates. See [guidance on configuring sampling][3].
4. Copy and run the installer command to load the Datadog RUM SDK Injector onto your Nginx module.

> 5. After the installer successfully installs the SDK Injector, restart your Nginx to begin collecting RUM sessions.

Reviewer: ```suggestion
5. After the installer successfully installs the SDK Injector, restart Nginx to begin collecting RUM sessions.
```
"your" is not grammatically wrong, just sounded a bit odd to me in this context.

{{% collapse-content title="Using the command line" level="h5" %}}

1. Run the Powershell command line as an administrator.
2. Set your Session and Session Replay sample rates. See [guidance on configuring sampling][1].
3. Copy and run the config command shown for each IIS site that you wish to inject RUM into.

[1]: /real_user_monitoring/guide/best-practices-for-rum-sampling/

{{% /collapse-content %}}

[1]: https://app.datadoghq.com/rum/list/create/

{{% /tab %}}
{{< /tabs >}}

## Updating your RUM application

You can adjust your Session Sampling and Session Replay Sampling rates from the Application Management page.

{{< tabs >}}
{{% tab "Nginx" %}}

To update your RUM Application:

1. Go to your RUM application from the [Application Management][1] list.
2. On the Instrument your application page, adjust the slider or enter a specific percentage in the input box for Session Sampling or Session Replay Sampling.
3. Copy and run the installer command in your `NGINX.conf` file.

[1]: https://app.datadoghq.com/rum/list

{{% /tab %}}

{{% tab "Windows IIS" %}}

To update your RUM Application:

1. Go to your RUM application from the [Application Management][1] list.

> 2. On the Instrument your application page, adjust the slider or enter a specific percentage in the input box for Session Sampling or Session Replay Sampling.

Reviewer: ```suggestion
2. On the Instrument Your Application page, adjust the slider or enter a specific percentage in the input box for Session Sampling or Session Replay Sampling.
```

> 3. Copy and replace the code in the Datadog RUM configs file for the IIS site that you instrumented.

Reviewer: ```suggestion
3. Copy and replace the code in the Datadog RUM config file for the IIS site that you instrumented.
```
Not sure about this one, feel free to ignore if I'm wrong!

[1]: https://app.datadoghq.com/rum/list

{{% /tab %}}
{{< /tabs >}}

## Limitations

The available functionality has the following important limitations:

- If serving compressed traffic, the Auto-Instrumentation method is not able to inject the JS scriptlet into the HTML traffic.
- This instrumentation method does not support any [advanced RUM configurations][3]. However, `allowedTracingUrls` and `excludedActivityUrls` are supported for Nginx web servers.

> - The SDK Injector does not inject into encrypted requests served by the Nginx or IIS related to TLS.

Reviewer: ```suggestion
- The SDK Injector does not inject into encrypted requests served by Nginx or IIS related to TLS.
```
